pkg/deprecatetemplates/allowlist.go

```diff
@@ -0,0 +1,150 @@
+package deprecatetemplates
+
+import (
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/ghodss/yaml"
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/config"
+
+	jc "github.com/openshift/ci-tools/pkg/jobconfig"
+)
+
+const (
+	presubmit  = "presubmit"
+	postsubmit = "postsubmit"
+	periodic   = "periodic"
+	release    = "release"
+	unknown    = "unknown"
+)
+
+func isGenerated(job config.JobBase) bool {
+	_, generated := job.Labels[jc.ProwJobLabelGenerated]
+	return generated
+}
+
+func getKind(job config.JobBase) string {
+	if strings.HasPrefix(job.Name, "pull-ci") {
+		return presubmit
+	} else if strings.HasPrefix(job.Name, "branch-ci-") {
+		return postsubmit
+	} else if strings.HasPrefix(job.Name, "release-") {
+		return release
+	} else if strings.HasPrefix(job.Name, "periodic-") {
+		return periodic
+	}
+
+	// this is fine, it is best effort
+	// the allowlist does not need to be 100% precise category-wise
+	return unknown
+}
+
+type blockedJob struct {
+	Generated bool   `json:"generated"`
+	Kind      string `json:"kind"`
+}
+
+type blockedJobs map[string]blockedJob
+
+func (b blockedJobs) Has(job config.JobBase) bool {
+	_, has := b[job.Name]
+	return has
+}
+
+func (b blockedJobs) Insert(job config.JobBase) {
+	b[job.Name] = blockedJob{
+		Generated: isGenerated(job),
+		Kind:      getKind(job),
+	}
+}
+
+func (b blockedJobs) Union(other blockedJobs) blockedJobs {
+	union := blockedJobs{}
+	for k, v := range b {
+		union[k] = v
+	}
+	for k, v := range other {
+		union[k] = v
+	}
+	return union
+}
+
+type deprecatedTemplateBlocker struct {
+	Description string      `json:"description"`
+	Jobs        blockedJobs `json:"jobs,omitempty"`
+}
+
+type deprecatedTemplate struct {
+	Name           string                               `json:"template_name"`
+	UnknownBlocker deprecatedTemplateBlocker            `json:"unknown_blocker"`
+	Blockers       map[string]deprecatedTemplateBlocker `json:"blockers,omitempty"`
+}
+
+func (d *deprecatedTemplate) insert(job config.JobBase) {
+	for _, blocker := range d.Blockers {
+		if blocker.Jobs.Has(job) {
+			return
+		}
+	}
+	d.UnknownBlocker.Jobs.Insert(job)
+}
+
+type Allowlist interface {
+	Insert(job config.JobBase, template string)
+	Save(path string) error
+}
+
+type allowlist struct {
+	Templates map[string]deprecatedTemplate `json:"templates"`
+}
+
+func (a *allowlist) Insert(job config.JobBase, template string) {
+	if a.Templates == nil {
+		a.Templates = map[string]deprecatedTemplate{}
+	}
+
+	if _, ok := a.Templates[template]; !ok {
+		a.Templates[template] = deprecatedTemplate{
+			Name: template,
+			UnknownBlocker: deprecatedTemplateBlocker{
+				Description: "unknown",
+				Jobs:        blockedJobs{},
+			},
+		}
+	}
+
+	dt := a.Templates[template]
```

Comment:
Nit: Why not just `a.Templates[template].insert(job)`?

Reply:
Hm, the reason was 
```
pkg/deprecatetemplates/allowlist.go:111:23: cannot call pointer method on a.Templates[template]
pkg/deprecatetemplates/allowlist.go:111:23: cannot take the address of a.Templates[template]
```

but it looks like I can just just make the `insert` have a non-pointer receiver.